gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,40 +313,105 @@ for _, ply in pairs(player.GetHumans()) do
     DrpCanHear[ply] = {}
 end
 
--- proxy function to take load from PlayerCanHearPlayersVoice, which is called a quadratic amount of times per tick,
--- causing a lagfest when there are many players
-local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    DrpCanHear[listener] = DrpCanHear[listener] or {}
-    if listener:IsBot() then return end
-    local shootPos = listener:GetShootPos()
-    for _, talker in ipairs(player.GetHumans()) do
-        local talkerShootPos = talker:GetShootPos()
-        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
-            (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
-                (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
-    end
-end
-hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
-    calcPlyCanHearPlayerVoice(ply)
-    if ply:IsBot() then return end
-    timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
-end)
+local gridSize = GM.Config.voiceDistance * 2 -- Grid cell size is equal to the size of the diamater of player talking
+local floor = math.floor -- Caching floor as we will need to use it a lot
+
+-- Grid based position check
+local grid
+local plyToGrid = {
+    {},
+    {}
+}
+
+
+timer.Create("DarkRPCanHearPlayersVoice", timeDelay, 0, function()
+    local players = player.GetHumans()
+
+
+    -- Clear old values
+    plyToGrid[0] = {}
+    plyToGrid[1] = {}
+    grid = {}
+
+    -- Get the grid position of every player O(N)
+    for _, ply in ipairs(players) do
+        local pos = ply:GetPos()
```

Comment:
As you calculate the player position at the beginning, you can just cache it. This will mean you can index it later instead of calling the position loads of times. The only thing I can think that is a downside is that it will be a bit outdated 
by only a minuscule amount.

Reply:
They could have teleported in that time. It should just stay as is.

Reply:
I presume you mean to cache it through each execution of this function. I believe that the entire concept works on the assumption are 'roughly' in the same area for 0.3+ seconds in which case, the minor change for caching positions is negligible.

Reply:
If the positions are cached at the start of the function, and used later in the same call tree, players can not be teleported. I presume @Kapppa is talking about the uses of GetPos on lines 373 and 393.